### PR TITLE
Make kernel-install --verbose more verbose and some autodetection logic fixups

### PR DESCRIPTION
**Pull request by keszybz**
I'm marking this as draft because I need to do some more testing and figure out how to handle upgrades from F34- which had a flat layout in /boot and why it stopped working now (https://bugzilla.redhat.com/show_bug.cgi?id=2071034).

### CODE REVIEW

**Comment by DaanDeMeyer** [comment]
All makes total sense to me. Only thing I'd consider is turning this pattern: 

```
[ "$KERNEL_INSTALL_VERBOSE" -gt 0 ] && [ -n "$initrd_generator" ] && \
    echo "$install_conf configures initrd_generator=$initrd_generator"
```

into a function maybe.

**Comment by keszybz** [comment]
> All makes total sense to me. Only thing I'd consider is turning this pattern:
> 
> ```
> [ "$KERNEL_INSTALL_VERBOSE" -gt 0 ] && [ -n "$initrd_generator" ] && \
>     echo "$install_conf configures initrd_generator=$initrd_generator"
> ```
> 
> into a function maybe.

I don't see how to do this easily… Bash functions are more like macros and don't provide good isolation. This form is at least easy to get right. With a function, we could abstract the check for `$KERNEL_INSTALL_VERBOSE`, but not the rest.

I think the last commit fixes the issue in Fedora. I'll ask for more testing tomorrow, or test this myself, but it's too late for me today.

**Comment by keszybz** [comment]
Y'all, please run `bootctl -v inspect` on your machines and check if the returned answers are no worse than before.

I can reproduce the original issue by upgrading [Fedora-Cloud-Base-34-1.2.x86_64.qcow2](https://download.fedoraproject.org/pub/fedora/linux/releases/34/Cloud/x86_64/images/Fedora-Cloud-Base-34-1.2.x86_64.qcow2) to F36 and `mkdir sudo mkdir /boot/efi/$(cat /etc/machine-id)` and installing a new kernel.

**Comment by keszybz** [comment]
I added a FIXME in the code, and another one is that `install.conf` hath no man page. But this isn't urgent.

File: src/kernel-install/kernel-install.in [hunk 1]
```diff
@@ -103,25 +103,55 @@ layout=
...
-    . /usr/lib/kernel/install.conf
+    install_conf="/usr/lib/kernel/install.conf"
+else
+    install_conf=
+fi
+
+if [ -n "$install_conf" ]; then
+    [ "$KERNEL_INSTALL_VERBOSE" -gt 0 ] && echo "Reading $install_conf…"
+    . "$install_conf"
+    # FIXME: This may override configuration in environment variables, e.g. $BOOT_ROOT.
```
Comment (behrmann) [inline src/kernel-install/kernel-install.in:116]: Wouldn't an (admittedly) ugly solution to this be
```suggestion
    # Sourcing $install_conf may override configuration in environment variables, e.g. $BOOT_ROOT.
    _BOOT_ROOT_FROMENV="$BOOT_ROOT"
    . "$install_conf"
    if [ -n "$_BOOT_ROOT_FROMENV" ] && [ "$BOOT_ROOT" != "$_BOOT_ROOT_FROMENV" ]; then
        BOOT_ROOT="$_BOOT_ROOT_FROMENV"
    fi
```
and so on for every other variable that should be thusly protected?

**Comment by DaanDeMeyer** [comment]
> Y'all, please run bootctl -v inspect on your machines and check if the returned answers are no worse than before.

You mean kernel-install -v inspect right? On a Fedora 36 machine (started at Fedora 34):

```
➜  systemd git:(kernel-install-verbose) build/kernel-install -v inspect
Reading /usr/lib/kernel/install.conf…
machine-id 9bc4fc1fd8c442b9817cd6fd77e03de6 acquired from /etc/machine-info
Entry-token candidates: 9bc4fc1fd8c442b9817cd6fd77e03de6 fedora Default
/efi/9bc4fc1fd8c442b9817cd6fd77e03de6 not found…
/efi/loader/entries not found…
/efi/fedora not found…
/efi/loader/entries not found…
/efi/Default not found…
/efi/loader/entries not found…
/boot/9bc4fc1fd8c442b9817cd6fd77e03de6 not found…
/boot/loader/entries exists, using BOOT_ROOT=/boot
No entry-token candidate matched, using "9bc4fc1fd8c442b9817cd6fd77e03de6" from machine-id
Entry-token directory not found, using layout=other
Using ENTRY_DIR_ABS=/boot/9bc4fc1fd8c442b9817cd6fd77e03de6/
Plugin files:
/usr/lib/kernel/install.d/10-devicetree.install
/usr/lib/kernel/install.d/20-grub.install
/usr/lib/kernel/install.d/20-grubby.install
/usr/lib/kernel/install.d/50-depmod.install
/usr/lib/kernel/install.d/50-dracut.install
/usr/lib/kernel/install.d/51-dracut-rescue.install
/usr/lib/kernel/install.d/60-kdump.install
/usr/lib/kernel/install.d/90-loaderentry.install
/usr/lib/kernel/install.d/92-crashkernel.install
/usr/lib/kernel/install.d/95-kernel-hooks.install
/usr/lib/kernel/install.d/99-grub-mkconfig.install
KERNEL_INSTALL_MACHINE_ID: 9bc4fc1fd8c442b9817cd6fd77e03de6
KERNEL_INSTALL_ENTRY_TOKEN: 9bc4fc1fd8c442b9817cd6fd77e03de6
KERNEL_INSTALL_BOOT_ROOT: /boot
KERNEL_INSTALL_LAYOUT: other
KERNEL_INSTALL_INITRD_GENERATOR: 
ENTRY_DIR_ABS: /boot/9bc4fc1fd8c442b9817cd6fd77e03de6/$KERNEL_VERSION
```

**Comment by keszybz** [comment]
> You mean kernel-install -v inspect right? On a Fedora 36 machine (started at Fedora 34):

Oops, indeed.
 
> ```
> ENTRY_DIR_ABS: /boot/9bc4fc1fd8c442b9817cd6fd77e03de6/$KERNEL_VERSION
> ```

This looks correct. I did some testing here, and I think this is doing the right thing. I'm fairly confident of the analysis of the reported issue, so I think we should just merge this and move on with the release.

File: src/kernel-install/kernel-install.in [hunk 2]
```diff
@@ -189,19 +189,16 @@ for suff in $ENTRY_TOKEN_SEARCH; do
...
-        BOOT_ROOT="$pref"
 
-        [ "$KERNEL_INSTALL_VERBOSE" -gt 0 ] && \
-            echo "$pref/loader/entries exists, using BOOT_ROOT=$BOOT_ROOT"
-        break
-    else
-        [ "$KERNEL_INSTALL_VERBOSE" -gt 0 ] && echo "$pref/loader/entries not found…"
-    fi
+        if [ -d "$pref/loader/entries" ]; then
+            BOOT_ROOT="$pref"
```
Comment (daandemeyer) [inline src/kernel-install/kernel-install.in]: Don't we need to conditionalize this on `[ -z "$BOOT_ROOT" ]` as well?

Comment (keszybz) [reply src/kernel-install/kernel-install.in]: Force pushed with this added.

Effectively it doesn't actually matter: if we get to this point and BOOT_ROOT is set, $pref must be the same, so we'd just reset the variable to the same value. But I think it's more elegant to have the check.

**Review by DaanDeMeyer — APPROVED** [review]

File: src/kernel-install/kernel-install.in [hunk 3]
```diff
@@ -226,6 +291,8 @@ PLUGINS="$(
 IFS="
 "
 
+[ "$KERNEL_INSTALL_VERBOSE" -gt 0 ] && echo -e "Plugin files:\n$PLUGINS"
```
Comment (bluca) [inline src/kernel-install/kernel-install.in:294]: @keszybz apparently `echo -e` is a bashism:

```
$ checkbashisms src/kernel-install/kernel-install.in 
possible bashism in src/kernel-install/kernel-install.in line 294 (echo -e):
[ "$KERNEL_INSTALL_VERBOSE" -gt 0 ] && echo -e "Plugin files:\n$PLUGINS"
```

File: src/kernel-install/kernel-install.in [hunk 4]
```diff
@@ -75,7 +75,7 @@ fi
 export KERNEL_INSTALL_VERBOSE=0
 if [ "$1" = "--verbose" ] || [ "$1" = "-v" ]; then
     shift
-    KERNEL_INSTALL_VERBOSE=1
+    export KERNEL_INSTALL_VERBOSE=1
```
Comment (poettering) [inline src/kernel-install/kernel-install.in:78]: this was dropped in 76b1274a5cb54acaa4a0f0c2e570d751f9067c06

Hmm, isn't the "export-and-set-in-one" thing a bashism? Shouldn't this be done in two lines?

/cc @nabijaczleweli

Comment (behrmann) [reply src/kernel-install/kernel-install.in:78]: shellcheck deems it clean and a quick check in with dash and busybox ash also support it.

Comment (mrc0mmand) [reply src/kernel-install/kernel-install.in:78]: It's not exactly bashism, it's even allowed by POSIX, but some shells don't support it. However, I can't recall any shell we "support" that doesn't support this syntax.

Comment (nabijaczleweli) [reply src/kernel-install/kernel-install.in:78]: I ran some tests on the more exotic/baseline shells I could think of (pdksh/mksh/yash/mrsh/NetBSD sh) and every one I tried supported this, so I don't think this is an issue for any shell in general.

**Pull request merged**